File: runtime/cpp/titan-ast-runtime-lib/RuntimeAutomataRichAstApplication.cpp
//
// Created by tian wei jun on 2024/10/26.
//

#include "RuntimeAutomataRichAstApplication.h"

RuntimeAutomataRichAstApplication::RuntimeAutomataRichAstApplication()
    : RuntimeAutomataAstApplication(),
      richResultConverter(AstGeneratorResult2RichResultConverter()) {}
RuntimeAutomataRichAstApplication::~RuntimeAutomataRichAstApplication() =
    default;

RuntimeAutomataAstApplication *RuntimeAutomataRichAstApplication::clone() {
  auto *app = new RuntimeAutomataRichAstApplication();
  this->cloneDataToCloner(app);
  // set newline
  app->richResultConverter.setNewline(this->richResultConverter.getNewline());
  return app;
}

void RuntimeAutomataRichAstApplication::setNewline(byte newline) {
  richResultConverter.setNewline(newline);
}

RichAstGeneratorResult *RuntimeAutomataRichAstApplication::buildRichAst(
    const std::string *sourceFilePath) {
  return richResultConverter.convert(buildAst(sourceFilePath));
}

File: runtime/java/titan-ast-runtime/src/titan/ast/runtime/RuntimeAutomataRichAstApplication.java
package titan.ast.runtime;

import java.io.InputStream;
import java.nio.charset.Charset;

/**
 * RuntimeAutomataAstApplication.
 *
 * @author tian wei jun
 */
public class RuntimeAutomataRichAstApplication extends RuntimeAutomataAstApplication
    implements Cloneable {
  AstGeneratorResult2RichResultConverter richResultConverter =
      new AstGeneratorResult2RichResultConverter();

  public void setNewline(byte newline) {
    richResultConverter.setNewline(newline);
  }

  public void setCharset(Charset charset) {
    richResultConverter.setCharset(charset);
  }

  public void setCharset(String charsetName) {
    richResultConverter.setCharset(charsetName);
  }

  public RichAstGeneratorResult buildRichAst(String sourceFilePath) {
    return richResultConverter.convert(buildAst(sourceFilePath));
  }

  public RichAstGeneratorResult buildRichAst(InputStream sourceByteInputStream) {
    return richResultConverter.convert(buildAst(sourceByteInputStream));
  }

  public RuntimeAutomataRichAstApplication clone() {
    RuntimeAutomataRichAstApplication app = null;
    app = (RuntimeAutomataRichAstApplication) super.clone();
    app.setContext(this.automataData);
    // set newline ,setCharset
    app.setNewline(this.richResultConverter.getNewline());
    app.setCharset(this.richResultConverter.getCharset());
    return app;
  }
}
